endpoint: remove Owner interface

mhofstetter opened the pull request:
This PR removes the endpoint regeneration `Owner` interface that is used by the endpoints logic to retrieve its dependencies.

The results is an intermediate state, where all the dependencies are explicitly on the functions that create endpoints.

As a next step, the "Endpoints API" (e.g. `EndpointManager`) should provide the functionality without requiring all the dependencies as parameters. These should be retrieved via Hive dependencies. This way we can gradually get rid of the dependency from Endpoint to the daemon. 

Please review the individual commits.

mhofstetter committed:
endpoint: remove components from endpoint regeneration.Owner interface

Currently, endpoint logic fetches most of its dependencies via the
`regeneration.Owner` interface directly from the daemon.

As a preparation to get rid of these dependencies on the daemon and
to introduce a proper Endpoint API that encapsulates the logic and
its required dependencies, this commit removes the following components
from the Owner interface to make the dependencies transparent.

* Loader
* Orchestrator
* CompilationLock
* BandwidthManager
* IPTablesManager
* IdentityManager

Signed-off-by: Marco Hofstetter <marco.hofstetter@isovalent.com>

mhofstetter committed:
endpoint: introduce EndpointBuildQueue

Currently, the endpoint build queue logic is implemented
on the daemon and the endpoint event handling logic retrieves
the instance via the Owner interface.

To remove the dependencies from the endpoint logic to the daemon,
this commit introduces a new Hive component `EndpointBuildQueue`.

This is just a preparation to eventually introduce a proper Endpoints
API that encapsulates the logic and dependencies. The queue is still
passed through via the daemon to the Endpoints.

Signed-off-by: Marco Hofstetter <marco.hofstetter@isovalent.com>

mhofstetter committed:
endpoint: remove SendNotification from regeneration Owner interface

This commit removes the SendNotification method from the Owner interface.

Endpoint logic should directly depend on the monitoragent to send events.

Signed-off-by: Marco Hofstetter <marco.hofstetter@isovalent.com>

mhofstetter committed:
endpoint: rename Owner to DNSRulesApi interface

This commit renames the Owner interface to `DNSRulesAPI` - as the
only remaining functions are DNS related.

Signed-off-by: Marco Hofstetter <marco.hofstetter@isovalent.com>

mhofstetter committed:
endpoint: remove implementations of DNSRulesAPI (ex. Owner)

This commit removes unnecessary test implementations of the new
`DNSRulesAPI` interface (ex. `Owner`).

Signed-off-by: Marco Hofstetter <marco.hofstetter@isovalent.com>

mhofstetter committed:
endpoint: remove policyRepoGetter and use policy.PolicyRepository directly

This commit removes the unnecessary `policyRepoGetter` interfaces and
replaces them with direct dependencies to the `PolicyRepository`.

Signed-off-by: Marco Hofstetter <marco.hofstetter@isovalent.com>

mhofstetter committed:
endpoint: cleanup owner reference test mocks

This commit removes no longer used owner reference test mocks.

Signed-off-by: Marco Hofstetter <marco.hofstetter@isovalent.com>

mhofstetter committed:
endpoint: remove DNSRulesAPI (ex. Owner) from EndpointManager.AddEndpoint

This commit removes the unused parameter `DNSRulesAPI` from the method
`EndpointManager.AddEndpoint`.

Signed-off-by: Marco Hofstetter <marco.hofstetter@isovalent.com>

mhofstetter committed:
fqdn: extract DNSRulesResolver from daemon

This commit extracts and moves the DNSRulesResolver logic into
a new Hive Component that is provided by the FQDN module.

Signed-off-by: Marco Hofstetter <marco.hofstetter@isovalent.com>

joamaki reviewed (approved):
Nice incremental step to the right direction!

christarazi commented on pkg/fqdn/endpoint/cell.go:
Nit: I think I'd just name this to be `pkg/fqdn/rules`, since `endpoint` isn't really that useful as a package name here. This package is about DNS rules.

mhofstetter replied:
good point - thanks! i renamed the package (`pkg/fqdn/rules`) and the component (`DNSRulesService`)

And naming and package layout will definitely be adjusted once we continue to modularize the FQDN package. This is probably not the final stage :sweat_smile:

christarazi reviewed (approved):
Awesome! 🧹

mhofstetter committed:
fqdn: introduce fqdn controlplane cell

With the introduction of the FQDN DNSRulesResolver cell, there are now
two FQDN related controlplane cells (including nameserver).

Therefore, this commit introduces a new FQDN Hive Cell that combines
the two.

Note: `DefaultDNS` (singleton) is an infrastructure cell.

Signed-off-by: Marco Hofstetter <marco.hofstetter@isovalent.com>

bimmlerd commented on cilium-health/launch/endpoint.go:
```diff
@@ -224,7 +223,7 @@ type EndpointAdder interface {
 // cleanup of prior cilium-health endpoint instances.
 func LaunchAsEndpoint(baseCtx context.Context,
-	owner regeneration.Owner,
+	dnsRulesApi endpoint.DNSRulesAPI,
```
nit: go style would be `dnsRulesAPI` - https://go.dev/wiki/CodeReviewComments#initialisms

applies throughout

mhofstetter replied:
right - thanks! if it's ok i will rename these occurrences in the follow up PR that will touch these functions again!

bimmlerd reviewed (approved):
tiny nit, thanks for the PR! Please keep on doing these 💯
